### PR TITLE
test/factory: pass testing.T around rather than errors for test fixtures

### PR DESCRIPTION
**Pull request by tychoish**

### CODE REVIEW

**Comment by creachadair** [comment]
I am sympathetic to the goal of this change, but I am not at all comfortable with hacking in a random extra parameter that is solely for testing. Is there some reason that I'm missing, why we should not make `MakeVote` et al. report an error? Either way we're having to update all the call sites anyway.

Or if we can't do that for some reason, I think I'd be happier with an "internal/external" split vs. a parameter.

**Comment by tychoish** [comment]
MakeVote can't just take a testing.T because it gets called in the e2e tests :/ so it's a bit of a kludge, I agree with your read though.

**Comment by creachadair** [comment]
> MakeVote can't just take a testing.T because it gets called in the e2e tests :/ so it's a bit of a kludge, I agree with your read though.

Sure, that is fine—but given that `MakeVote` already reports an error, could we address this by returning an error instead of panicking for signature failure? We could inject a sentinel error so the test can check for it, e.g., 

```go
var ErrSignatureFailed = errors.New("vote signature failed")

// ...
err := val.SignVote(ctx, chainID, vpb)
if err != nil {
   return nil, ErrSignatureFailed
}
v.Signature = vpb.Signature
return v, nil

// ...
v, err := factory.MakeVote(/* blah */)
if errors.Is(err, factory.ErrSignatureFailed) {
   t.Fatal(err)
}
// ...
```

or words to that effect?

**Comment by tychoish** [comment]
I fixed up makevote as you suggested. No sentinel error, because I think the current tests aren't that fine grained, but will look at other error handling cases too.

**Comment by tychoish** [comment]
removed the error handler pattern entirely.

File: internal/test/factory/block.go [hunk 1]
```diff
@@ -89,13 +85,12 @@ func MakeHeader(h *types.Header) (*types.Header, error) {
...
-	if err != nil {
-		panic(err)
-	}
 	return h
 }
+
+func MakeRandomHeader(t *testing.T) *types.Header {
+	t.Helper()
+	return MakeHeader(t, &types.Header{})
+}
```
Comment (creachadair) [inline internal/test/factory/block.go:96]: This probably deserves documentation if we're going to keep it. That said: I recommend we omit it entirely, use `MakeHeader(t, new(types.Header))` directly, and update the documentation there. The randomness is implemented there already anyway.

Say, for example:

```go
// MakeHeader fills the unset fields of h in-place with random legal values, so
// that the resulting header passes basic validation checks. It returns h.
```

or words to that effect.

File: internal/test/factory/vote.go [hunk 2]
```diff
@@ -2,12 +2,29 @@ package factory
...
 	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
 	"github.com/tendermint/tendermint/types"
 )
 
+// ErrorHandler checks an error and returns true if the error is
+// non-nil. Pass an ErrorHandler to a function in a case where you
+// might want to use a *testing.T but cannot because one is not
+// available (e.g. as is the case in some e2e test fixtures and in
+// TestMain.)
+type ErrorHandler func(error) bool
```
Comment (creachadair) [inline internal/test/factory/vote.go:18]: I don't think anything checks the return value of this. Can we drop the `bool` entirely and have this be a simple error logger?

Comment (tychoish) [reply internal/test/factory/vote.go:18]: oh I thought I deleted this entirely

Comment (creachadair) [reply internal/test/factory/vote.go:18]: I was going to suggest that, but it's still being used.

File: internal/test/factory/block.go (hunk 1 above)
Comment (tychoish) [reply internal/test/factory/block.go]: I removed it, as it's only used in one place and is kind of silly.

**Review by creachadair — APPROVED** [review]

**Pull request merged**